toolchain/check/context.cpp

```diff
@@ -505,6 +532,102 @@ auto Context::ImplicitAsRequired(Parse::Node parse_node, SemIR::NodeId value_id,
+      context.semantics_ir().GetTypeBlock(tuple_elem_types_id);
+
+  if (tuple_elem_types.size() !=
+      context.semantics_ir().GetArrayBoundValue(array_bound_id)) {
+    return false;
+  }
+
+  // TODO: We should do a proper convertibility check here for each
+  // tuple element.
+  return llvm::all_of(tuple_elem_types,
+                      [element_type_id = element_type_id](auto type_id) {
+                        return type_id == element_type_id;
+                      });
+}
+
+// Performs a conversion from a tuple to an array type.
+static auto ConvertTupleToArray(Context& context, SemIR::Node tuple_type,
+                                SemIR::Node array_type,
+                                SemIR::TypeId array_type_id,
+                                SemIR::NodeId value_id) -> SemIR::NodeId {
+  auto [array_bound_id, element_type_id] = array_type.GetAsArrayType();
+  auto tuple_elem_types_id = tuple_type.GetAsTupleType();
+  const auto& tuple_elem_types =
+      context.semantics_ir().GetTypeBlock(tuple_elem_types_id);
+
+  auto array_bound = context.semantics_ir().GetArrayBoundValue(array_bound_id);
+  if (tuple_elem_types.size() != array_bound) {
+    CARBON_DIAGNOSTIC(
+        ArrayInitArgCountMismatch, Error,
+        "Cannot initialize array of {0} element(s) from {1} initializer(s).",
+        decltype(array_bound), size_t);
+    context.emitter().Emit(
+        context.semantics_ir().GetNode(value_id).parse_node(),
+        ArrayInitArgCountMismatch, array_bound, tuple_elem_types.size());
+    return SemIR::NodeId::BuiltinError;
+  }
+
+  auto refs_id = context.semantics_ir().AddNodeBlock();
+  auto& refs = context.semantics_ir().GetNodeBlock(refs_id);
+  llvm::ArrayRef<SemIR::NodeId> literal_elems;
+
+  // If we're initializing from a tuple literal, use its elements directly.
+  // Otherwise, materialize a temporary if needed and index into the result.
+  auto value = context.semantics_ir().GetNode(value_id);
+  while (value.kind() == SemIR::NodeKind::StubReference) {
+    value_id = value.GetAsStubReference();
+    value = context.semantics_ir().GetNode(value_id);
+  }
+  if (value.kind() == SemIR::NodeKind::TupleValue) {
+    literal_elems =
+        context.semantics_ir().GetNodeBlock(value.GetAsTupleValue());
+  } else {
+    value_id = context.MaterializeIfInitializing(value_id);
+  }
+
+  // Arrays are always initialized in-place. Tentatively allocate a temporary
+  // as the destination for the array initialization.
+  auto return_slot_id = context.AddNode(
+      SemIR::Node::TemporaryStorage::Make(value.parse_node(), array_type_id));
+
+  // Initialize each element of the array from the corresponding element of the
+  // tuple.
+  for (auto [i, src_type_id] : llvm::enumerate(tuple_elem_types)) {
+    // TODO: Add a new node kind for indexing an array at a constant index
+    // so that we don't need an integer literal node here.
+    auto int_id = context.AddNode(SemIR::Node::IntegerLiteral::Make(
```

Comment:
Can you rename this to index_id or something? Having int_id and init_id next to each other feels close to a typo...

Reply:
Done.